api/public/public_api/endpoints/device_activate.py

```diff
@@ -32,64 +32,98 @@
 from schematics import Model
 from schematics.types import StringType
 
-from selene.api import PublicEndpoint
-from selene.api import generate_device_login
+from selene.api import generate_device_login, PublicEndpoint
+from selene.api.pantacor import get_pantacor_device
 from selene.data.device import DeviceRepository
 from selene.util.cache import DEVICE_PAIRING_TOKEN_KEY
 
 
 class ActivationRequest(Model):
+    """Data model of the fields in the request."""
+
     token = StringType(required=True)
     state = StringType(required=True)
-    platform = StringType(default='unknown')
-    coreVersion = StringType(default='unknown')
-    enclosureVersion = StringType(default='unknown')
+    platform = StringType(default="unknown")
+    core_version = StringType(default="unknown")
+    enclosure_version = StringType(default="unknown")
     platform_build = StringType()
+    pantacor_device_id = StringType()
 
 
 class DeviceActivateEndpoint(PublicEndpoint):
+    """API endpoint for activating a device, the last step in device pairing."""
+
+    _device_repository = None
+
+    @property
+    def device_repository(self):
+        """Lazily load an instance of the device repository"""
+        if self._device_repository is None:
+            self._device_repository = DeviceRepository(self.db)
+
+        return self._device_repository
+
     def post(self):
+        """Process a HTTP POST request."""
         activation_request = self._validate_request()
         pairing_session = self._get_pairing_session()
         if pairing_session is not None:
-            device_id = pairing_session['uuid']
+            device_id = pairing_session["uuid"]
             self._activate(device_id, activation_request)
-            response = (
-                generate_device_login(device_id, self.cache),
-                HTTPStatus.OK
-            )
+            response = (generate_device_login(device_id, self.cache), HTTPStatus.OK)
         else:
-            response = '', HTTPStatus.NOT_FOUND
+            response = "", HTTPStatus.NOT_FOUND
+
         return response
 
-    def _validate_request(self):
+    def _validate_request(self) -> dict:
+        """Validate the contents of the API request against the data model."""
         activation_request = ActivationRequest(self.request.json)
         activation_request.validate()
 
-        return activation_request
+        return activation_request.to_native()
 
     def _get_pairing_session(self):
         """Get the pairing session from the cache.
 
         The request must have same state value as that stored in the
         pairing session.
         """
-        pairing_session_token = self.request.json['token']
+        pairing_session_token = self.request.json["token"]
```

Comment:
I can't see where `self.request` gets assigned. Is there any chance it will be mutated during execution? Would it be better to pass it in as an argument?

Reply:
The request object is part of the flask API.  The base Selene endpoint class inherits from Flask's MethodView class, which supplies the request object.